stetho/src/main/java/com/facebook/stetho/inspector/network/BasePrettyPrinterFactory.java

```diff
@@ -0,0 +1,35 @@
+package com.facebook.stetho.inspector.network;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintWriter;
+
+import com.facebook.stetho.common.Util;
+import com.facebook.stetho.inspector.network.AsyncPrettyPrinter;
+import com.facebook.stetho.inspector.network.AsyncPrettyPrinterFactory;
+
+/**
+ * Implementation of {@link AsyncPrettyPrinterFactory} which generates asynchronous pretty printers
+ * that allows callers to pretty print responses with specific header names
+ */
+public abstract class BasePrettyPrinterFactory implements AsyncPrettyPrinterFactory {
+
+  @Override
+  public AsyncPrettyPrinter getInstance(String headerName, String headerValue) {
+    //TODO: async download schema
+    final String schema = "";
+    return new AsyncPrettyPrinter() {
+      //TODO: pool byte array if this method is used frequently to avoid GC churn
+      public void printTo(PrintWriter output, InputStream payload) throws IOException {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        Util.copy(payload, out, new byte[1024]);
+        output.write(doPrint(out.toByteArray(), schema));
+        output.close();
+      }
+    };
+  }
+
+  protected abstract String doPrint(byte[] payload, String schema);
```

Comment:
This dismantles the optimization I asked for.  Push the efficient interface all the way to the final implementor.